workflows/validate.py

```diff
@@ -0,0 +1,78 @@
+"""
+Runs the json validator on a json file and raises an exception if there is an error
+Intended to be used as a hook for repositories which are developing recipes
+
+Example of how this could be used in a .pre-commit-config.yaml file:
+```
+# Validate json recipes
+- repo: https://github.com/DiamondLightSource/python-workflows
+  hooks:
+  - name: Recipe validation
+    id: recipe-validation
+    language: python
+    files: ^recipes/
+    entry: workflows-recipe-validate
+```
+"""
+
+import logging
+import workflows.recipe
+import workflows
+import sys
+import json
+import argparse
+
+
+def validate_recipe(json_filename):
+    """Reads a json file, tries to turn it into a recipe and then validates it.
+    Exits on exception with non-zero error"""
+
+    # Read in the file
+    try:
+        with open(json_filename) as f:
+            recipe_text = f.read()
+    except Exception as e:
+        logging.exception("Could not read recipe from {0}".format(json_filename))
+        raise e
+
+    # Turn it into a recipe and validate
+    try:
+        # Create a recipe object and validate it
+        workflows.recipe.Recipe(recipe_text).validate()
+    except json.decoder.JSONDecodeError as e:
+        logging.error(
+            "JSON error in recipe {0}, please address this".format(json_filename)
+        )
+        logging.error("{0} at line {1} col {2}".format(e.msg, e.lineno, e.colno))
+        raise e
+    except workflows.Error as e:
+        logging.error(
+            "Problem in error in recipe {0}, please address this".format(json_filename)
+        )
+        logging.error("{0}".format(e))
+        raise e
+    except Exception as e:
+        logging.error("Problem in recipe {0} please address this".format(json_filename))
+        logging.error("{0}".format(e))
+        raise e
+
+
+def main():
+    """Run the program from entry point"""
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "files", nargs="+", help="file or list of files to be validated"
+    )
+    args = parser.parse_args()
+
+    # Validate each file in turn
+    for file in args.files:
+        try:
+            validate_recipe(file)
+        except Exception:
+            sys.exit(1)
```

Comment:
I give you 5 files to validate. You stop after file 1. Probably more useful to validate all, then exit with error if one of them failed.
You eat up the exceptions here anyway, so why not eat them up in `validate_recipe` and just return `True/False`? Then you can rewrite this as
```python
if not all(validate_recipe(f) for f in args.files):
  sys.exit(1)
```

Reply:
Yes it would perhaps be better to validate all and only return errors on what failed so that you have a list of all the problems. It used to eat all the exceptions in validate_recipe, however this made it harder to test for specific exceptions in the unittests, hence dragging it out to the main function.

Reply:
This section has now increased a bit in size but retains useful features as command line tool while allowing the validate_recipe function to continue separately.
Tries to work a bit more like black, exits silently if no errors but otherwise lists the erroneous files and their issues.